src/cluster.c

```diff
@@ -1820,6 +1862,102 @@ void clusterUpdateSlotsConfigWith(clusterNode *sender, uint64_t senderConfigEpoc
     }
 }
 
+/* Cluster ping extensions.
+ *
+ * The ping/pong/meet messages support arbitrary extensions to add additional
+ * metadata to the messages that are sent between the various nodes in the
+ * cluster. The extensions take the form:
+ * [ Header length + type (8 bytes) ] 
+ * [ Extension information (Arbitrary length, but must be 8 byte padded) ]
+ */
+
+
+/* Returns the length of a given extension */
+static uint16_t getPingExtLength(clusterMsgPingExt *ext) {
+    return ntohl(ext->length);
+}
+
+/* Returns true if the provided extension starts at the end of 
+ * the cluster message, indicating we are done parsing extensions */
+static int isTerminalPingExt(clusterMsgPingExt *ext, clusterMsg *hdr) {
+    return (((char *) hdr) + ntohl(hdr->totlen)) == (char *) ext;
+}
+
+/* Returns the initial position of ping extensions */
+static clusterMsgPingExt *getInitialPingExt(clusterMsg *hdr, uint16_t count) {
+    clusterMsgPingExt *initial = (clusterMsgPingExt*) &(hdr->data.ping.gossip[count]);
+    if (isTerminalPingExt(initial, hdr)) {
+        return NULL;
+    }
+    return initial;
+} 
+
+/* Given a current ping extension, returns the start of the next extension */
+static clusterMsgPingExt *getNextPingExt(clusterMsgPingExt *ext, clusterMsg *hdr) {
+    clusterMsgPingExt *next = (clusterMsgPingExt *) (((char *) ext) + getPingExtLength(ext));
+    if (isTerminalPingExt(next, hdr)) {
+        return NULL;
+    }
+    return next;
+}
+
+/* Returns the exact size needed to store the hostname. The returned value
+ * will be 8 byte padded. */
+int getHostnamePingExtSize() {
+    /* If hostname is not set, we don't send this extension */
+    if (!myself->hostname) return 0;
+
+    int totlen = sizeof(clusterMsgPingExt) + EIGHT_BYTE_ALIGN(strlen(myself->hostname) + 1);
+    return totlen;
+}
+
+/* Write the hostname ping extension at the start of the cursor. This function
+ * will update the cursor to point to the end of the written extension and
+ * will return the amount of bytes written. */
+int writeHostnamePingExt(clusterMsgPingExt **cursor) {
+    /* If hostname is not set, we don't send this extension */
+    if (!myself->hostname) return 0;
+
+    /* Add the hostname information at the extension cursor */
+    clusterMsgPingExtHostname *ext = &(*cursor)->ext[0].hostname;
+    size_t hostname_len = strlen(myself->hostname);
+    memcpy(ext->hostname, myself->hostname, hostname_len);
```

Comment:
Should we be more defensive and not assume the target is already zero-filled?

Reply:
We do a zcalloc earlier, so I think it's safe to assume it's zero filled.

Reply:
It just makes this code a bit more fragile in case it ends up in different code paths in the future.